Hide lambda body squashing under newlines.source=fold

poslegm opened the pull request:
Should fix #1826 for default behavior

Lambdas don't squash with default config, but «braces to parens» works with `fold`:
```
<<< squash
newlines.source=fold
rewrite.rules=[RedundantBraces]
===
files.test("basic") { name =>
  assertEquals(name, "basic-setup")
}
>>>
files.test("basic")(name => assertEquals(name, "basic-setup"))
```

#### Next steps

- squash blocks with `fold`:
```
<<< failing now
newlines.source=fold
===
files.test("basic") {
  assertEquals(name, "basic-setup")
}
>>>
files.test("basic") { assertEquals(name, "basic-setup") }
```
- note at the website documentation that entire power of `RedundantBraces` can be accessed only with `newlines.source=fold`

@kitbellew what do you think?

Also I really like the idea of keeping newlines configuration in universal formatting style sets like `fold`, `unfold`, etc. Now we have a lot of granular configs like `alwaysBeforeCurlyBraceLambdaParams` which affects formatting only in particular cases. They conflict each other. What about to gradually deprecate them and move its functionality into `newlines.source`? It may be related not only to `newlines`: I think that highly specialized settings from entire scalafmt config can be combined into several enums.

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:
```diff
@@ -1080,13 +1081,19 @@ class FormatOps(val tree: Tree, val initStyle: ScalafmtConfig) {
-  )(implicit style: ScalafmtConfig): (Boolean, NewlineT) =
+  )(implicit style: ScalafmtConfig): (Boolean, NewlineT) = {
+    val tryToSquashLambdaInOneLine =
+      style.newlines.sourceIs(Newlines.fold) || newlines == 0
```
the problem here is that you are looking at `newlines` when source is `unfold`, and we can't do that. `sourceIgnored` is better.

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:
```diff
@@ -1080,13 +1081,19 @@ class FormatOps(val tree: Tree, val initStyle: ScalafmtConfig) {
       case NewlineCurlyLambda.preserve =>
-        (newlines == 0, if (newlines >= 2) Newline2x else Newline)
+        (
+          tryToSquashLambdaInOneLine,
```
this redefines the meaning of `preserve` and might end squashing when the user had a newline (or a couple!).

kitbellew commented:
@poslegm i think i understand what's happening. `afterCurlyLambdas` was designed to control _blank lines_ (i.e., multiple consecutive newlines) and not line breaks. `never` means no blank lines, `always` means one blank line, and `preserve` is keep as-is. however, none of them was designed to achieve no newlines.

perhaps i shouldn't have used this parameter to achieve lambda folding in 2.4.2, but the cat is out of the bag. @olafurpg you requested the change, what's your opinion? is it reasonable to undo the fix to #1027 introduced in 2.4.2 or force them to reformat all of their code using `newlines.source=fold`?

another option is to add a new label to `newlines.afterCurlyLambdas`, it shouldn't increase the complexity or change interaction with other parameters.

poslegm commented:
Another case to afterCurlyLambdas may be good workaround. I will take a
look tomorrow.

ср, 1 апр. 2020 г., 3:09 Albert Meltzer <notifications@github.com>:

> @poslegm <https://github.com/poslegm> i think i understand what's
> happening. afterCurlyLambdas was designed to control *blank lines* (i.e.,
> multiple consecutive newlines) and not line breaks. never means no blank
> lines, always means one blank line, and preserve is keep as-is. however,
> none of them was designed to achieve no newlines.
>
> perhaps i shouldn't have used this parameter to achieve lambda folding in
> 2.4.2, but the cat is out of the bag. @olafurpg
> <https://github.com/olafurpg> you requested the change, what's your
> opinion? is it reasonable to undo the fix to #1027
> <https://github.com/scalameta/scalafmt/issues/1027> introduced in 2.4.2
> or force them to reformat all of their code using newlines.source=fold?
>
> another option is to add a new label to newlines.afterCurlyLambdas, it
> shouldn't increase the complexity or change interaction with other
> parameters.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scalameta/scalafmt/pull/1832#issuecomment-606952470>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDYJSXQXYKRCTQV4WQE6WDRKKA4NANCNFSM4LYEIWFQ>
> .
>

kitbellew commented:
> Another case to afterCurlyLambdas may be good workaround. I will take a look tomorrow.
> ср, 1 апр. 2020 г., 3:09 Albert Meltzer <notifications@github.com>:

надеюсь, не первоапрельская шутка? :)

olafurpg commented:
> another option is to add a new label to newlines.afterCurlyLambdas, it shouldn't increase the complexity or change interaction with other parameters.
 
I’m not sure about the implications of this but I think we can change the default without breaking changes. Users who upgraded and like the current behavior should be able to change some settings to preserve the current behavior.

poslegm commented:
> надеюсь, не первоапрельская шутка? :)

время сейчас не весёлое

poslegm commented:
Ok, I reworked it a bit.

Introduced `newlines.afterCurlyLambda=squash`. It will squash lambda body as you might have guessed. And `newlines.afterCurlyLambda=never` shouldn't unfold already one-liners.

So new default behavior is

```
<<< stay as is
xs.map { x =>
  x + 1
}
>>>
xs.map { x =>
  x + 1
}
```

```
<<< stay as is
xs.map { x => x + 1 }
>>>
xs.map { x => x + 1 }
```

Users who want to squash lambdas now don't need to enable full `source=fold`: `newlines.afterCurlyLambda=squash` will be enough.

And please take a look on new restrictions for `newline.source` combinations with `newlines.afterCurlyLambda`.

poslegm commented on scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:
```diff
@@ -1082,7 +1083,8 @@ class FormatOps(val tree: Tree, val initStyle: ScalafmtConfig) {
     style.newlines.afterCurlyLambda match {
-      case NewlineCurlyLambda.never => (true, Newline)
+      case NewlineCurlyLambda.squash => (true, Newline)
+      case NewlineCurlyLambda.never => (newlines == 0, Newline)
```
Hmmm

Are you sure that `never` should unfold lambda body like this?

```scala
<<< ONLY meh
xs.map { x => x + 1 }
>>>
xs.map { x =>
  x + 1
}
```

In my mind lambda with _no one newline_ meets the requirements of `newlines.afterCurlyLambda=never`

kitbellew replied:
@poslegm, you are not allowed to check newlines with `fold`/`unfold`. but your example makes sense, so you could change `never` case to something like
```
  case NewlineCurlyLambda.never =>
    val space = style.newlines.source match {
      case Newlines.fold => true
      case Newlines.unfold => false
      case _ => newlines == 0
    }
    (space, Newline)
```

poslegm replied:
Understand now. Thank you!

poslegm commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala:
```diff
@@ -168,6 +168,12 @@ case class ScalafmtConfig(
       if (optIn.configStyleArguments && align.openParenDefnSite)
         errors += s"optIn.configStyleArguments with align.openParenDefnSite"
     }
+    if (newlines.source == Newlines.fold && newlines.afterCurlyLambda != NewlineCurlyLambda.squash) {
```
removed it
